Alternatives to tooltips docs (#238)
primer/design: content/accessibility/tooltip-alternatives.mdx

## Guidelines 

### For designers

- Reserve tooltips for components like icon buttons. 
- Keep your tooltip text minimal.
- Only include tooltips on other components as a last resort.
- **Never** include tooltips on non-interactive components (`div`, `span`, `p`).  

<DoDontContainer>
   <Do>
      <img width="456" alt="Icon buttons with tooltips for edit, favorite, and sponsor" src="https://user-images.githubusercontent.com/40274682/160182083-a97ba63f-a297-4613-bb52-d4a8ef10cc10.png"/>
      <Caption>Use a tooltip to give a text label to an icon button</Caption>
   </Do>
   <Dont>
      <img width="456" alt="Non-interactive text element activating a long and wordy tooltip on hover" src="https://user-images.githubusercontent.com/40274682/160182084-948a7969-9c84-4c08-8801-4d8a5253b1ce.png"/>
      <Caption>Don't use tooltips on non-interactive elements</Caption>
   </Dont>
</DoDontContainer>

<Box mb={3} display="flex" alignItems="flex-start" alignItems="center" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <Box as="p" mt="0">

>        <h4>Annotations</h4>If you have an appropriate tooltip in your design, annotate the tooltip carefully with the preferred type (is it a label or description?) to ensure proper semantics.</Box>

Reviewer: ```suggestion
       <h4>Annotations</h4>If you have an appropriate tooltip in your design, annotate the tooltip carefully with the preferred type (whether it's a label or description) to ensure proper semantics.</Box>
```

    <img
        width="456"
        alt="Mock of an icon button and its tooltip with an annotation pointing to the tooltip with 'label' text"
        src="https://user-images.githubusercontent.com/40274682/160183578-4f66e628-c207-47a4-8520-00a37e26def4.png"
    />
</Box>

### For developers

If you come across a design or a page with a tooltip on a static element (for example, `div`, `span`, `p`), remove it. Consult a designer for an alternative way of conveying the information.

> If you determine that tooltips are appropriate for your use case, make sure to carefully follow guidelines set in [PVC: Tooltip component](https://primer.style/view-components/components/alpha/tooltip).

Reviewer: ```suggestion
If you determine that tooltips are appropriate for your use case, make sure to carefully follow guidelines set in [Primer ViewComponents: Tooltip component](https://primer.style/view-components/components/alpha/tooltip).
```
Let's not use the acronyms.

## Alternatives to tooltips

### 1. Persist the content

If possible, persist the content so it's always available rather than using a tooltip, which hides content by default. This ensures that the content is always discovered and surfaced to users regardless of device.

<DoDontContainer>
   <Do>
      <img width="456" alt="Information about how many discussions can be pinned added to the confirmation dialog" src="https://user-images.githubusercontent.com/40274682/160192785-13764229-7c42-4fa9-b752-b2ec741d4d4a.png"/>
      <Caption>Find an appropriate permanent place to display additional information.</Caption>
   </Do>
   <Dont>
      <img width="456" alt="Information about how many discussions can be pinned within tooltip on non-interactive element" src="https://user-images.githubusercontent.com/40274682/160192791-2d69905a-bd1f-4508-9c0a-9e968b6d4ac3.png"/>

>       <Caption>Don't use a tooltip to add in additional helpful infomation</Caption>

Reviewer: ```suggestion
      <Caption>Don't use a tooltip to add additional helpful information</Caption>
```

### 2. Don't duplicate content

If the tooltip duplicates content that is already available on the page, remove it.

<DoDontContainer>
   <Do>
      <img width="456" alt="Pointer hovering over a link with no tooltip" src="https://user-images.githubusercontent.com/40274682/160201665-0f3bf527-fef7-4d79-b81b-af194385be5d.png"/>
      <Caption>Remove any duplicate tooltip text.</Caption>
   </Do>
    <Dont>
       <img width="456" alt="Screenshot of tooltip with GitHub username, @inkblotty, on a static span that says @inkblotty" src="https://user-images.githubusercontent.com/40274682/160201667-d90e076d-7441-45f4-b0b6-620693921105.png"/>
       <Caption>Don't set tooltips that duplicate the trigger element content.</Caption>
    </Dont>
</DoDontContainer>

### 3. Use a modal

Consider using a modal, which is accessible for mobile users and allows you to structure content that may otherwise be crammed into a tooltip.

<DoDontContainer>
    <Do>
       <img width="456" alt="Screenshot of GitHub Actions page with a modal that conveys workflow file information about a runner" src="https://user-images.githubusercontent.com/40274682/160210527-2319e408-d433-407f-aac2-66eaf9e63246.png" />

>        <Caption>Do use a pattern that is accessible a larger number of users such as a modal</Caption>

Reviewer: ```suggestion
       <Caption>Do use a pattern that is accessible to a larger number of users such as a modal</Caption>
```

---
title: Tooltip alternatives
---

## Overview

Tooltips are often used to convey information. However, tooltips are rarely appropriate, and its misuse can result in a myriad of accessibility issues.
Always consider **not** using a tooltip for an improved user experience.

## Why?

- Tooltips are hidden by default making it easy to miss, so it should never be used to convey critical information.
- Tooltips are never accessible on mobile devices.

> - Tooltips should **never** be set on static elements. Tooltips on static elements are not accessible to keyboard users and screen reader users.

Reviewer 2: I think the term is intended to encompass elements not included in the HTML spec’s list of [interactive elements](https://html.spec.whatwg.org/multipage/interactive-elements.html).

## Guidelines 

### For designers

Only include tooltips on your design as a last resort. Never include tooltips on static elements. 

> If you have an appropriate tooltip in your design, be sure to annotate the tooltip carefully with the preferred type (for example, label or description) to ensure proper semantics.

Author: Maybe "an appropriate tooltip usecase" is more appropriate

> If you have an appropriate tooltip in your design, be sure to annotate the tooltip carefully with the preferred type (for example, label or description) to ensure proper semantics.

Reviewer: is it more clear to say "a correct tooltip use case"?

> If you have an appropriate tooltip in your design, be sure to annotate the tooltip carefully with the preferred type (for example, label or description) to ensure proper semantics.

Reviewer: unless we want it to be more fuzzy

### For developers

If you come across a design or a page with a tooltip on a static element, remove it. Consult a designer for an alternative way of conveying the information.

> If you determine that tooltips are appropriate for your usecase, make sure to carefully follow guidelines set in [PVC: Tooltip component](https://primer.style/view-components/components/alpha/tooltip).

Author: I'm not aware of where React is at with the Tooltip but if there's an accessible tooltip we can link to, happy to add it here

## What are alternatives to tooltips?

### 1. Always persist the content

> If possible, persist the content so it's available rather than using a tooltip which hides content by default. This ensures that the content is always discovered and surfaced to consumers regardless of device.

Reviewer 3: One example that might be relevant is the "Pin Issue" button in an issue sidebar describing the max number of issues that could be pinned.

It could be reworked to say "Pin issue (max 3)" or something.

### 2. Don't show content

If the tooltips duplicate content that is already available on the page, remove it.

<DoDontContainer stacked>
    <Dont>

>        <Caption>Don't set tooltips that duplicate the trigger element content, nor on static trigger elements</Caption>

Author: There are several violations in each of these examples. Would it be worth acknowledging them?

>        <Caption>Don't set tooltips that duplicate the trigger element content, nor on static trigger elements</Caption>

Reviewer: I guess so, if it doesn't become super long?

>        <Caption>Don't set tooltips that duplicate the trigger element content, nor on static trigger elements</Caption>

Author: This has been resolved with @ichelsea latest updates!

>        <Caption>Don't set tooltips that duplicate the trigger element content, nor on static trigger elements</Caption>

Author: Ready for review :)

    </Dont>
</DoDontContainer>

### 4. Use a summary disclosure

Consider using a summary disclosure which will be accessible for mobile users and offers ability to show or hide content similar to a tooltip.

<DoDontContainer stacked>
    <Do>
       <Caption>Do use a pattern that is accessible for more groups of users such as a summary disclosure</Caption>
       <img width="800" alt="Screenshot of GitHub pricing page that uses a summary disclosure to hide and show information for point made" src="https://user-images.githubusercontent.com/16447748/159560761-382819b2-99a8-4f59-87ba-0447fe68e59e.png" />
    </Do>
    <Dont>
       <Caption>Don't set tooltips on static elements nor set long content inside a tooltip</Caption>
       <img width="800" alt="Screenshot of GitHub pricing page that initially used a tooltip to hide and show long descriptions for each point made" src="https://user-images.githubusercontent.com/16447748/159561347-1c2c9d38-7082-44f7-b9d8-642c8cca8085.png" />

>     </Dont>

Author: Hmmm I think it's helpful to see the before after state so I'd like to keep this here

>     </Dont>

Reviewer: I don't see this part anymore, guessing it was edited?
